File: C++/10891.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

int dfn[100001], vnum(0);
vector<int> V[100001];

int dfs(int p, int mp = -1) {
	dfn[p] = ++vnum;
	int res = 0;
	for (int q : V[p]) {
		if (q == mp)
			continue;
		if (!dfn[q]) {
			int tmp = dfs(q, p);
			if (res && tmp) {
				cout << "Not cactus";
				exit(0);
			}
			else if (tmp)
				res = tmp;
		}
		else if (dfn[p] > dfn[q]){
			if (res) {
				cout << "Not cactus";
				exit(0);
			}
			res = q;
		}
	}
	return ((res == p) ? 0 : res);
}


int main() {
	int N, M;
	cin >> N >> M;
	for (int i = 0; i < M; i++) {
		int x, y;
		cin >> x >> y;
		V[x].push_back(y);
		V[y].push_back(x);
	}
	dfs(1);
	cout << "Cactus";
}

File: Python/10891.py
import sys
sys.setrecursionlimit(100001)


def dfs(p, mp=-1):
    global vnum
    dfn[p] = vnum
    vnum += 1
    res = 0
    for q in V[p]:
        if q == mp:
            continue
        if not dfn[q]:
            tmp = dfs(q, p)
            if res and tmp:
                print('Not cactus')
                sys.exit(0)
            elif tmp:
                res = tmp
        elif dfn[p] > dfn[q]:
            if res:
                print('Not cactus')
                sys.exit(0)
            res = q
    return 0 if res == p else res


N, M = map(int, sys.stdin.readline().split())
dfn = [0] * (N + 1)
vnum = 1
V = [[] for _ in range(N + 1)]
for _ in range(M):
    x, y = map(int, sys.stdin.readline().split())
    V[x].append(y)
    V[y].append(x)
dfs(1)
print('Cactus')
